ksqldb-execution/src/main/java/io/confluent/ksql/execution/interpreter/InterpretedExpression.java

```diff
@@ -0,0 +1,753 @@
+import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.KsqlException;
+import io.confluent.ksql.util.Pair;
+import java.math.BigDecimal;
+import java.math.MathContext;
+import java.math.RoundingMode;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.Struct;
+
+@SuppressWarnings("checkstyle:ClassDataAbstractionCoupling")
+public class InterpretedExpression implements ExpressionEvaluator {
+
+  private final ExpressionTypeManager expressionTypeManager;
+  private final FunctionRegistry functionRegistry;
+  private final LogicalSchema schema;
+  private final KsqlConfig ksqlConfig;
+  private final Expression expression;
+  private final SqlType expressionType;
+
+  public InterpretedExpression(final FunctionRegistry functionRegistry,
+      final LogicalSchema schema,
+      final KsqlConfig ksqlConfig,
+      final Expression expression,
+      final SqlType expressionType,
+      final ExpressionTypeManager expressionTypeManager) {
+    this.ksqlConfig = ksqlConfig;
+    this.expression = expression;
+    this.expressionType = expressionType;
+    this.expressionTypeManager = expressionTypeManager;
+    this.functionRegistry = functionRegistry;
+    this.schema = schema;
+  }
+
+  @VisibleForTesting
+  Object evaluate(final GenericRow row) {
+    final Pair<Object, SqlType> evaluator =
+        new Evaluator(row).process(expression, null);
+    return evaluator.getLeft();
+  }
+
+  @Override
+  public Object evaluate(
+      final GenericRow row,
+      final Object defaultValue,
+      final ProcessingLogger logger,
+      final Supplier<String> errorMsg
+  ) {
+    try {
+      return evaluate(row);
+    } catch (final KsqlException e) {
+      // This is likely something that would have been caught previously at compile time, so allow
+      // it to bubble up.
+      throw e;
+    } catch (final Exception e) {
+      logger.error(RecordProcessingError.recordProcessingError(errorMsg.get(), e, row));
+      return defaultValue;
+    }
+  }
+
+  @Override
+  public Expression getExpression() {
+    return expression;
+  }
+
+  public SqlType getExpressionType() {
+    return expressionType;
+  }
+
+  private class Evaluator implements ExpressionVisitor<Pair<Object, SqlType>, Void> {
+
+    private final GenericRow row;
+
+    Evaluator(final GenericRow row) {
+      this.row = row;
+    }
+
+    private Pair<Object, SqlType> visitIllegalState(final Expression expression) {
+      throw new IllegalStateException(
+          format("expression type %s should never be visited", expression.getClass()));
+    }
+
+    private Pair<Object, SqlType> visitUnsupported(final Expression expression) {
+      throw new UnsupportedOperationException(
+          format(
+              "not yet implemented: %s.visit%s",
```

Comment:
if this is ever hit, should we suggest people to disable the interpreter config?

Reply:
We could, though it should be noted that the compiler also has limitations and a similar set of things that will throw this exception.  (If done correctly, they are in fact the same set of limitations.)

So, if this is triggered, it's likely and hopeful that the same thing wouldn't work for the compiler.  That said, maybe people won't be good at adding functionality to both at once and the compiler will always be ahead?  I could try to ensure this by writing tests for unsupported features in the parity test.  What do you think?

Reply:
> the compiler also has limitations

Ah, so this PR fully implements the compiled functionality? In that case why not default it to true for pull queries?

> I could try to ensure they do both by writing tests for unsupported features in the parity test. 

That would definitely be a good stop-gap measure. We might even want to consider removing the Janino codepath altogether for persistent queries (though obviously that's a longer follow-up) if the performance is only marginally worse (over time we can improve the performance of this).

This also gives me an idea, it would be interesting to test this PR by running all the QTT tests, and setting persistent queries to use this evaluator. Would everything work?